Manifold/Inference.swift

```diff
@@ -1,10 +1,11 @@
 //  Copyright (c) 2015 Rob Rix. All rights reserved.
 
-public typealias ConstraintSet = Multiset<Constraint>
-
 /// Infers the type, assumptions, and constraints for a given `expression`.
 public func infer(expression: Expression) -> (Type, assumptions: AssumptionSet, constraints: ConstraintSet) {
 	return expression.analysis(
+		ifConstant: { c in
+			(c.type, assumptions: [:], constraints: [])
```

Comment:
Hypothetical type-annotated expressions (like Swift’s `as` family of operators) would lead to very similar constraint/assumption construction.